File: elixir/series/series.exs
defmodule StringSeries do
  @doc """
  Given a string `s` and a positive integer `size`, return all substrings
  of that size. If `size` is greater than the length of `s`, or less than 1,
  return an empty list.
  """
  @spec slices(s :: String.t(), size :: integer) :: list(String.t())
  def slices(s, size) do
    series(s, size, [])
  end

  defp series(s, size, acc) do
    cond do
      size < 1 -> acc
      size > String.length(s) -> acc
      true -> series(String.slice(s, 1..-1), size, acc ++ [String.slice(s, 0..size-1)])
    end
  end
end

File: go/series/series.go
package series

const testVersion = 2

func All(n int, s string)(out []string){
	for i := 0; (i + n) <= len(s); i++ {
		out = append(out, s[i:(i+n)])
	}
	return
}

func UnsafeFirst(n int, s string)(out string){
	out = s[0:n]
	return
}

func First(n int, s string)(first string, ok bool){
	if n > len(s) {
		ok = false
		return
	} else {
		ok = true
	}
	first = UnsafeFirst(n, s)
	return
}
